File: src/2025-03-04-03-28-40--deepseek-chat-0304/go/exercises/practice/bowling/bowling.go
package bowling

import "errors"

type Game struct {
	rolls []int
}

func NewGame() *Game {
	return &Game{
		rolls: make([]int, 0, 21), // Max 21 rolls in a game
	}
}

func (g *Game) Roll(pins int) error {
	if pins < 0 || pins > 10 {
		return ErrInvalidPins
	}
	
	// Check for invalid roll in frame (except 10th)
	if len(g.rolls)%2 == 1 && len(g.rolls) < 18 {
		if g.rolls[len(g.rolls)-1] + pins > 10 {
			return ErrInvalidPins
		}
	}
	
	g.rolls = append(g.rolls, pins)
	return nil
}

func (g *Game) Score() (int, error) {
	if len(g.rolls) < 12 || len(g.rolls) > 21 {
		return 0, ErrIncompleteGame
	}
	
	score := 0
	rollIndex := 0
	
	for frame := 0; frame < 10; frame++ {
		if isStrike(g.rolls, rollIndex) {
			score += 10 + strikeBonus(g.rolls, rollIndex)
			rollIndex++
		} else if isSpare(g.rolls, rollIndex) {
			score += 10 + spareBonus(g.rolls, rollIndex)
			rollIndex += 2
		} else {
			score += sumOfBallsInFrame(g.rolls, rollIndex)
			rollIndex += 2
		}
	}
	
	return score, nil
}

// Helper functions
func isStrike(rolls []int, rollIndex int) bool {
	return rolls[rollIndex] == 10
}

func isSpare(rolls []int, rollIndex int) bool {
	return rolls[rollIndex]+rolls[rollIndex+1] == 10
}

func strikeBonus(rolls []int, rollIndex int) int {
	return rolls[rollIndex+1] + rolls[rollIndex+2]
}

func spareBonus(rolls []int, rollIndex int) int {
	return rolls[rollIndex+2]
}

func sumOfBallsInFrame(rolls []int, rollIndex int) int {
	return rolls[rollIndex] + rolls[rollIndex+1]
}

// Errors
var (
	ErrInvalidPins    = errors.New("invalid number of pins")
	ErrIncompleteGame = errors.New("game is not complete")
)

File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/bowling/bowling.py
class BowlingGame:
    def __init__(self):
        self.rolls = []

    def roll(self, pins):
        if pins < 0 or pins > 10:
            raise ValueError("Invalid number of pins")
            
        # Check for invalid frame totals (except 10th frame)
        if len(self.rolls) % 2 == 1 and self.rolls[-1] + pins > 10 and len(self.rolls) < 18:
            raise ValueError("Frame total cannot exceed 10")
            
        self.rolls.append(pins)
        
        # Special 10th frame validation
        if len(self.rolls) >= 19:
            # After first roll in 10th frame
            if len(self.rolls) == 19 and self.rolls[18] != 10 and self.rolls[18] + pins > 10:
                raise ValueError("Frame total cannot exceed 10")
            # After second roll in 10th frame if no strike/spare
            if len(self.rolls) == 20 and sum(self.rolls[18:20]) < 10:
                raise IndexError("Cannot roll after game is complete")

    def score(self):
        if len(self.rolls) < 20:
            raise IndexError("Game is not complete")
        if len(self.rolls) == 20 and sum(self.rolls[18:20]) >= 10:
            raise IndexError("Game is not complete")
            
        total = 0
        roll_index = 0
        
        for frame in range(10):
            if self.is_strike(roll_index):
                total += 10 + self.strike_bonus(roll_index)
                roll_index += 1
            elif self.is_spare(roll_index):
                total += 10 + self.spare_bonus(roll_index)
                roll_index += 2
            else:
                total += self.sum_of_balls_in_frame(roll_index)
                roll_index += 2
                
        return total

    def is_strike(self, roll_index):
        return self.rolls[roll_index] == 10

    def is_spare(self, roll_index):
        return self.rolls[roll_index] + self.rolls[roll_index + 1] == 10

    def sum_of_balls_in_frame(self, roll_index):
        return self.rolls[roll_index] + self.rolls[roll_index + 1]

    def strike_bonus(self, roll_index):
        return self.rolls[roll_index + 1] + self.rolls[roll_index + 2]

    def spare_bonus(self, roll_index):
        return self.rolls[roll_index + 2]
